src/envoy/auth/http_filter.cc

```diff
@@ -51,5 +115,90 @@ void JwtVerificationFilter::setDecoderFilterCallbacks(
   decoder_callbacks_ = &callbacks;
 }
 
+void JwtVerificationFilter::ReceivePubkey(HeaderMap& headers,
+                                          std::string issuer_name, bool succeed,
+                                          const std::string& pubkey) {
+  auto iss_it = calling_issuers_.find(issuer_name);
+  auto& iss = iss_it->second;
+  iss->failed_ = !succeed;
+  if (succeed) {
+    iss->pkey_ = pubkey;
+  }
+  iss->loaded_ = true;
+  calling_issuers_.erase(iss_it);
+
+  // if receive all responses, proceed to verification
+  if (calling_issuers_.empty()) {
+    CompleteVerification(headers);
+  }
+}
+
+void JwtVerificationFilter::CompleteVerification(HeaderMap& headers) {
+  const HeaderEntry* entry = headers.get(kAuthorizationHeaderKey);
+  if (entry) {
+    const HeaderString& value = entry->value();
+    if (strncmp(value.c_str(), kAuthorizationHeaderTokenPrefix.c_str(),
+                kAuthorizationHeaderTokenPrefix.length()) == 0) {
+      std::string jwt(value.c_str() + kAuthorizationHeaderTokenPrefix.length());
+
+      for (const auto& iss : config_->issuers_) {
+        if (iss->failed_) {
+          continue;
+        }
+
+        /*
+         * TODO: update according to change of JWT lib interface
+         */
+        // verifying and decoding JWT
+        std::unique_ptr<rapidjson::Document> payload;
+        if (iss->pkey_type_ == "pem") {
+          payload = Auth::Jwt::Decode(jwt, iss->pkey_);
+        } else if (iss->pkey_type_ == "jwks") {
+          payload = Auth::Jwt::DecodeWithJwk(jwt, iss->pkey_);
+        }
+
+        if (payload) {
+          // verification succeeded
+          auto payload_str = JsonToString(payload.get());
+
+          // Check the issuer's name.
+          if (payload->HasMember("iss") && (*payload)["iss"].IsString() &&
+              (*payload)["iss"].GetString() == iss->name_) {
+            /*
+             * TODO: check exp claim
+             */
+
+            /*
+             * TODO: replace appropriately
+             */
+            headers.addReferenceKey(AuthorizedHeaderKey(), payload_str);
+
+            // Remove JWT from headers.
+            headers.remove(kAuthorizationHeaderKey);
```

Comment:
why add the header and delete it right away?

Reply:
First one is adding a new header("Istio-Auth-UserInfo"), and second one is deleting Authorization header
(I guess you thought `AuthorizedHeaderKey()` and `kAuthorizationHeaderKey` are the same)